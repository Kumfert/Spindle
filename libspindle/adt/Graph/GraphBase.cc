//
//
//

#include "spindle/GraphBase.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

bool 
GraphBase::resize( const int newNVtxs, const int newNEdges, const int newNBoundVtxs=0, const int newNCutEdges=0 ) {
  
  adjHead.resize( newNVtxs + newNBoundVtxs + 1 );
  adjSelf.resize( newNVtxs + newNBoundVtxs );
  adjList.resize( newNEdges + newNCutEdges );
  if ( vtxWeight.notNull() ) { 
    vtxWeight.resize( newNVtxs );
  }
  if ( edgeWeight.notNull() ) { 
    edgeWeight.resize( 2 * newNEdges );
  }
  
  nVtxs = newNVtxs;
  nEdges = newNEdges;
  nBoundVtxs = newNBoundVtxs;
  nNonZeros = newNCutEdges + 2*newNEdges;

  return true;
}

int
GraphBase::deg( const int i ) const {
  if ( ( i>=0 ) && (i<nVtxs) ) {
    if ( vtxWeight.size() == 0 ) {
      return ( adjHead[i+1] - adjHead[i] );
    } else { 
      int temp = 0;
      for (int j = adjHead[i]; j<adjHead[i+1]; ++j) {
	if ( adjList[j] != i ) {
	  temp += vtxWeight[ adjList[ j ] ];
	}
      }
      return temp;
    }
  } else {
    return -1;
  }
}
