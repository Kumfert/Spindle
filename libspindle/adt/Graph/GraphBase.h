//
// GraphBase.h
//
//
//  Graphs are structurally symmetric, but not necessarily numerically
//  symmetric.  In the latter case, edge weights are neccessary if for
//  nothing else than to hold ones and zeros
//
//  Furthermore, graphs can be weighted
//
//  Furthermore, graphs can be partitioned.  
//  A partitioned graph stores the list of cut edges and vertices.
//
//  Not all graph algorithms are guaranteed to recognize a graph with boundary
//  vertices
//
//            nVtxs nBndryVtxs
//           +-----+---+
//           |     |   |
//           |     |   |
//           |     |   |
//           +-----+---+
//
#ifndef SPINDLE_GRAPH_BASE_H_
#define SPINDLE_GRAPH_BASE_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

SPINDLE_BEGIN_NAMESPACE
//class Graph;
class GraphBase {
  // friend class Graph;
protected:
  int nVtxs;      // number of vertices in graph... not including boundary vtxs
  int nEdges;     // number of edges in the graph ... not including cut edges
  int nBoundVtxs; // number of boundary vertices
  int nNonZeros;

  SharedArray<int> adjHead;   // length = nVtxs + nBoundVtxs + 1;
  SharedArray<int> adjSelf;   // length = nVtxs + nBoundVtxs;
  SharedArray<int> adjList;   // length = 2 (nEdges + nCutEdges) 

  SharedArray<int> vtxWeight; // length = nVtxs or zero
  SharedArray<int> edgeWeight;// length = 2 * nEdges or zero
  
  GraphBase() { } // protected default constructor

public:

  bool resize( const int newNVtxs, const int newNEdges, const int newNBoundVtxs=0, const int newNCutEdges=0 );

  int queryNVtxs() const { return nVtxs; }
  int queryNEdges() const { return nEdges; }
  int queryNNonZeros() const { return 2*nEdges; }
  int queryNBoundVtxs() const { return nBoundVtxs; }
  int queryNCutEdges() const { return adjList.size() - 2*nEdges; }

  int size() const { return  nVtxs; }
  int deg(const int i) const;
 
  //data structure access
  const SharedArray<int>& getAdjHead() const { return adjHead; }
  const SharedArray<int>& getAdjList() const { return adjList; }
  const SharedArray<int>& getAdjSelf() const { return adjSelf; } 
  const SharedArray<int>& getVtxWeight() const { return vtxWeight; }
  const SharedArray<int>& getEdgeWeight() const { return edgeWeight; }
};

SPINDLE_END_NAMESPACE

#endif
