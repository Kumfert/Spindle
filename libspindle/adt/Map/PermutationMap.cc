//
// PermutationMap.cc
//
//  $Id: PermutationMap.cc,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is povided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  The PermutationMap class is meant to abstract away the concept of
//  a permutation vector which may be of old2New or new2old varieties.
//
//  A PermutationMap is not in a valid state until finalize() is called
//  to insure that both old2New and new2Old are consistent with each
//  other.
//

#include "spindle/PermutationMap.h"

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

#ifdef __FUNC__
#undef __FUNC__
#endif

PermutationMap::PermutationMap() { 
  incrementInstanceCount( PermutationMap::MetaData );
  currentState = EMPTY;
}

void
PermutationMap::validate() {
  if ( currentState == VALID ) { return; }
  int *src; 
  int *dest;
  if ( changedMostRecent == NEITHER ) {
    src = old2New.begin();
    dest = new2Old.begin();
    if ((*src >= 0)&&(*dest <= 0)) { 
      changedMostRecent = OLD2NEW; 
    } else if ((*src <= 0)&&(*dest >= 0)) { 
      changedMostRecent = NEW2OLD; 
      src = new2Old.begin();
      dest = old2New.begin();
    }
  } else if ( changedMostRecent == OLD2NEW ) { 
    src = old2New.begin();
    new2Old.init( 0 );
    dest = new2Old.begin();
  } else { // if ( changedMostRecent == NEW2OLD ) {
    src = new2Old.begin();
    old2New.init( 0 );
    dest = old2New.begin();
  }
  {for (int i=0; i<sz; i++) { 
    int perm_i = src[i];
    if ( perm_i == -1 ) { continue; }
    if ( ( perm_i < 0) ||  ( perm_i >= sz ) ) { 
      currentState = INVALID;
      return; 
    }
    dest[perm_i]++;
  }}
  // ... now make sure each vertex only occurred once
  int * p = dest;
  { for (int i=0; i<sz; i++) { 
    if ( ( *p != 1 ) && (*p != 0) ) { 
      currentState = INVALID;
      return;
    }
  }}

  // ... finally copy the reverse permutation 
  {for (int i=0; i<sz; i++) { 
    dest[ src[ i ] ] = i; 
  }}
  currentState = VALID;
  changedMostRecent = NEITHER;
  return;
}

bool
PermutationMap::reset() {
  old2New.init( -1 );
  new2Old.init( -1 );
  currentState = EMPTY;
  changedMostRecent = NEITHER;
  return true;
}


const SharedArray<int>& 
PermutationMap::getOld2New() const {
  return old2New;
}

SharedArray<int>& 
PermutationMap::getOld2New() {
  if ( (!isValid() ) && (changedMostRecent == NEW2OLD) ) {
    validate();
    assert( isValid() );
  }
  currentState = UNKNOWN;
  changedMostRecent = OLD2NEW;
  return old2New;
}

const SharedArray<int>& 
PermutationMap::getNew2Old() const {
  return new2Old;
}

SharedArray<int>& 
PermutationMap::getNew2Old() {
  if ((!isValid()) && (changedMostRecent == OLD2NEW) ) {
    validate();
    assert( isValid() );
  }
  currentState = UNKNOWN;
  changedMostRecent = NEW2OLD;
  return new2Old;
}

void 
PermutationMap::resize( const int n ) { 
  currentState = EMPTY;
  changedMostRecent = NEITHER;
  sz = n;
  old2New.resize(n);
  new2Old.resize(n);
}

void
PermutationMap::dump( FILE * fp ) const {
  if ( isValid() ) {
    fprintf(fp,"PermutationMap:");
    const int* p = getNew2Old().lend();
    for( int i=0; i<size(); i++) {
      fprintf(fp,"%d ", p[i] );
    }
  }
}

void
PermutationMap::prettyPrint( FILE * fp, bool new2old_perm, bool printLength, 
			     bool onesOffset ) const { 

  if ( !isValid() ) {
    return ;
  }
  int offset = (onesOffset) ? 1 : 0 ;
  
  if ( printLength == true ) {
    fprintf( fp,"%d\n", size() );
  }
  
  const int * p;
  if ( new2old_perm ) { 
    p = new2Old.lend();
  } else { 
    p = old2New.lend();
  }

  for(const int* cur = p, *stop=p + size(); cur <stop; ++cur ) {
    fprintf( fp, "%d\n", *cur + offset);
  }
} 

void
PermutationMap::storeObject( SpindleArchive& ar ) const {
  if ( isValid() ) {
    const int *p = getNew2Old().lend();
    size_t my_size = size();
    pair< const int*, size_t> my_pair( p,my_size );
    ar << my_pair;
  } else {
    ar << -1;
  }
}
#define __FUNC__ "PermutationMap::loadObject( SpindleArchive& ar )"
void
PermutationMap::loadObject( SpindleArchive& ar ) {
  size_t new_size = ar.peekLength();
  if ( (int)new_size == -1 ) { 
    int junk;
    ar >> junk;
    currentState = INVALID;
  } else {
    resize(new_size);
    int* p = getNew2Old().begin();
    pair<int*, size_t> new_pair( p, new_size );
    
    ar >> new_pair;
    
    WARN_IF( new_pair.second != new_size, "Warning, expected size=%d, but extracted only %d.", 
	     new_pair.second, new_size );
    changedMostRecent = NEW2OLD;
    validate();
  }
}
#undef __FUNC__

SPINDLE_IMPLEMENT_PERSISTANT( PermutationMap, SpindlePersistant )


