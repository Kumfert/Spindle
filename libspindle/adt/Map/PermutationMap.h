//
//  PermutationMap.h -- an abstraction of a permutation vector.
//
//  $Id: PermutationMap.h,v 1.2 2000/02/18 01:31:49 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  The PermutationMap class is meant to abstract away the concept of
//  a permutation vector which may be of old2New or new2old varieties.
//
//  A PermutationMap is not in a valid state until validate() is called
//  to insure that both old2New and new2Old are consistent with each
//  other.
//

#ifndef SPINDLE_PERMUTATION_H_
#define SPINDLE_PERMUTATION_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class PermutationMap : public SpindlePersistant {
  SPINDLE_DECLARE_PERSISTANT( PermutationMap )

private:
  PermutationMap();
  SharedArray<int> old2New;
  SharedArray<int> new2Old;

  enum { NEITHER, OLD2NEW, NEW2OLD } changedMostRecent;
  int sz;
public:
  // constructors/destructors
  PermutationMap( const int n );
  PermutationMap( const PermutationMap& perm );
  virtual ~PermutationMap();

  int size() const;

  // access specific permuation vectors
  const SharedArray<int>& getOld2New() const;
  SharedArray<int>& getOld2New();

  const SharedArray<int>& getNew2Old() const;
  SharedArray<int>& getNew2Old();


  virtual void validate();
  virtual void dump( FILE * stream ) const;

  virtual bool reset();
  void resize( const int n );
  void prettyPrint( FILE * fp, bool new2old_perm=true, bool printLength=false, 
		    bool onesOffset=false ) const; 

};

inline 
int PermutationMap::size() const {
  return sz;
}

inline 
PermutationMap::PermutationMap( const int n )
  : old2New(n), new2Old(n), sz(n) {
    incrementInstanceCount( PermutationMap::MetaData );
    reset();
}

inline 
PermutationMap::PermutationMap( const PermutationMap& perm ) 
  : sz( perm.size() ) {
    incrementInstanceCount( PermutationMap::MetaData );
    old2New.import( perm.old2New );
    new2Old.import( perm.new2Old );
    changedMostRecent = perm.changedMostRecent;
    currentState = perm.currentState;
}

inline 
PermutationMap::~PermutationMap() {
  decrementInstanceCount( PermutationMap::MetaData );
}

SPINDLE_END_NAMESPACE

#endif


