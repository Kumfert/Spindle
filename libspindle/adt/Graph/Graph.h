//
// Graph.h
//
// $Id: Graph.h,v 1.2 2000/02/18 01:31:45 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997-1999, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef SPINDLE_GRAPH_H_
#define SPINDLE_GRAPH_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_GRAPH_BASE_H_
#include "spindle/GraphBase.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

#ifdef MATLAB_MEX_FILE
extern "C" { 
#include "mex.h"
} // end extern "C"
#endif

SPINDLE_BEGIN_NAMESPACE

class MatrixBase;

class Graph: public GraphBase, public SpindlePersistant { 

  SPINDLE_DECLARE_PERSISTANT( Graph )
  
public:
  typedef const int* const_iterator;

  //constructors/destructors
  Graph();
  Graph( const int NVtxs, int *AdjHead, int *AdjList ); //assume ownership of arrays, unweighted
  Graph( const int NVtxs, const int *AdjHead, const int *AdjList ); // do not own arrays, unweighted
  Graph( const int NVtxs, int *AdjHead, int *AdjList, int *VtxWeights, int* EdgeWeights );
  Graph( const int NVtxs, const int *AdjHead, const int *AdjList, const int *VtxWeights, const int* EdgeWeights );
  Graph( GraphBase* inputGraph );
  Graph( const GraphBase* inputGraph );
  Graph( MatrixBase* inputMatrix );
  Graph( const MatrixBase* inputMatrix );
#ifdef MATLAB_MEX_FILE
  Graph( const mxArray* mat ); // constructor on matlab's Matrix
#endif 
  virtual ~Graph();

  //iterate over adjacency list of vertex i
  const_iterator begin_adj( int i ) const { return  (isValid()) ? (adjList.begin() + adjHead[i]) : 0 ; }
  const_iterator   end_adj( int i ) const { return  (isValid()) ? (adjList.begin() + adjHead[i+1]) : 0 ; }
  
  //iterate over low adjacency list of vertex i
  const_iterator begin_ladj( int i ) const { return  (isValid()) ? (adjList.begin() + adjHead[i]) : 0 ; }
  const_iterator   end_ladj( int i ) const { return  (isValid()) ? (adjList.begin() + adjSelf[i]) : 0 ; }

  //iterate over high adjacency list of vertex i
  const_iterator begin_hadj( int i ) const { return  (isValid()) ? (adjList.begin() + adjSelf[i]) : 0 ; }
  const_iterator   end_hadj( int i ) const { return  (isValid()) ? (adjList.begin() + adjHead[i+1]) : 0 ; }

  // required for setting variables
  virtual void validate();
  virtual void dump( FILE * fp ) const ;
  virtual bool reset();
  void prettyPrint( FILE * fp ) const ;
  //  bool isValid() const { return isValid(); } defined by SpindlePersistant

  const SharedArray<int>& getAdjHead()  const{ return adjHead; }
  const SharedArray<int>& getAdjList()  const{ return adjList; }
  const SharedArray<int>& getAdjSelf()  const{ return adjSelf; } 
  const SharedArray<int>& getVtxWeight()  const{ return vtxWeight; }
  const SharedArray<int>& getEdgeWeight()  const{ return edgeWeight; }

  SharedArray<int>& getAdjHead() { currentState=UNKNOWN; return adjHead; }
  SharedArray<int>& getAdjList() { currentState=UNKNOWN; return adjList; }
  SharedArray<int>& getAdjSelf() { currentState=UNKNOWN; return adjSelf; } 
  SharedArray<int>& getVtxWeight() { currentState=UNKNOWN; return vtxWeight; }
  SharedArray<int>& getEdgeWeight() { currentState=UNKNOWN; return edgeWeight; }

  bool setGhostNodes( const int nBoundVtxs, const int nCutEdges );
};

SPINDLE_END_NAMESPACE

#endif
