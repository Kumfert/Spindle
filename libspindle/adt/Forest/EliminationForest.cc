//
// EliminationForest.cc
//
//  $Id: EliminationForest.cc,v 1.2 2000/02/18 01:31:44 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//

#include "spindle/EliminationForest.h"

#ifndef SPINDLE_PERMUTATION_H_
#include "spindle/PermutationMap.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

SPINDLE_IMPLEMENT_PERSISTANT( EliminationForest, GenericForest )

EliminationForest::EliminationForest() {
  g = 0;
  currentState = EMPTY;
}

EliminationForest::EliminationForest( const Graph* graph ) {
  incrementInstanceCount( EliminationForest::MetaData );
  g = graph;
  // set a null parent and ancestor
  const int n = g->size();
  SharedArray<int> ancestor( n );
  ancestor.init(-1);
  parent.resize(0);
  GenericForest::resize(n);
  parent.resize(n);
  parent.init(-1);
  
  for( int i=0; i<n; ++i) {
    for( const int* jj=g->begin_adj(i), *stop_jj = g->end_adj(i); 
	 jj < stop_jj; ++jj ) {
      int j = *jj;
      if ( j < i ) { // ii is an ancestor of jj
	// find the root of the tree containing the node `jj'
	while( (ancestor[j]!=-1) & (ancestor[j]!=i) ) {
	  int temp = ancestor[j];
	  ancestor[j] = i; 
	  j = temp;
	}
	// make node `ii' the root of this subtree, if not already true
	if ( ancestor[j] == -1 ) {
	  ancestor[j] = i;
	  parent[j] = i;
	}
      } // end if `ii' is ancestor of `jj'
    } // end for all jj adj to ii
  } // end for all ii
  currentState = UNKNOWN;
  createChildSibling();
  createDoubleLink();
}


EliminationForest::EliminationForest( const Graph* graph, const SharedPtr<PermutationMap>& perm ) {
  incrementInstanceCount( EliminationForest::MetaData );
  g = graph;
  
  // set a null parent and ancestor
  const int n = g->size();
  SharedArray<int> ancestor( n );
  ancestor.init(-1);
  GenericForest::resize(n);
  parent.resize(n);
  parent.init(-1);
  
  const int* old2new = perm->getOld2New().lend();
  const int* new2old = perm->getNew2Old().lend();
  
  // programmer's note:  `ii' & `jj' refer to the vertices in the new ordering
  //                      `i' & `j' refer to the same vertices in the old ordering
  
  for( int ii=0; ii<n; ++ii) {
    int i = new2old[ii];
    for( const int* jp=g->begin_adj(i), *stop_jp = g->end_adj(i); 
	 jp < stop_jp; ++jp ) {
      int j = *jp;
      int jj = old2new[ j ];
      if ( jj < ii ) { // ii is an ancestor of jj
	/* old code.  put the tree in terms of new indices, not old
	// find the root of the tree containing the node `jj'
	while( (ancestor[jj]!=-1) & (ancestor[jj]!=ii) ) {
	  int temp = ancestor[jj];
	  ancestor[jj] = ii; 
	  jj = temp;
	}
	// make node `ii' the root of this subtree, if not already true
	if ( ancestor[jj] == -1 ) {
	  ancestor[jj] = ii;
	  parent[jj] = ii;
	}
	*/
	// find the root of the tree containing the node `j'
	while( (ancestor[j]!=-1) & (ancestor[j]!=i) ) {
	  int temp = ancestor[j];
	  ancestor[j] = i; 
	  j = temp;
	}
	// make node `i' the root of this subtree, if not already true
	if ( ancestor[j] == -1 ) {
	  ancestor[j] = i;
	  parent[j] = i;
	}
      } // end if `ii' is ancestor of `jj'
    } // end for all jj adj to ii
  } // end for all ii
  currentState = UNKNOWN;
  createChildSibling();
  createDoubleLink();
}// end


EliminationForest::EliminationForest( const Graph* graph, const int n, const int* Parent )
  : GenericForest( n, Parent ) {
    incrementInstanceCount( EliminationForest::MetaData );
    g = graph;
}

EliminationForest::~EliminationForest() {
  decrementInstanceCount( EliminationForest::MetaData );
}

#define __FUNC__ "void EliminationForest::loadObject( SpindleArchive & ar )"
void
EliminationForest::loadObject( SpindleArchive & ar ) {
  ERROR( SPINDLE_ERROR_UNIMP_FUNC , "Method not yet implemented.");
  if ( &ar );
}
#undef __FUNC__

#define __FUNC__ "void EliminationForest::loadObject( SpindleArchive & ar )"
void
EliminationForest::storeObject( SpindleArchive& ar ) const {
  ERROR( SPINDLE_ERROR_UNIMP_FUNC , "Method not yet implemented.");
  if( &ar ) ;
}
#undef __FUNC__

  
