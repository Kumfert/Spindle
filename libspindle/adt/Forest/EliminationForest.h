//
// EliminationForest.h
//
//  $Id: EliminationForest.h,v 1.2 2000/02/18 01:31:44 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//

#ifndef SPINDLE_ARRAY_E_FOREST_H
#define SPINDLE_ARRAY_E_FOREST_H_

#ifndef SPINDLE_AUTO_PTR_H_
#include "spindle/SharedPtr.h"
#endif

#ifndef SPINDLE_ARRAY_FOREST_H_
#include "spindle/GenericForest.h"
#endif 

#ifndef SPINDLE_GRAPH_H_
#include "spindle/Graph.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class PermutationMap;

class EliminationForest : protected GenericForest { 

  SPINDLE_DECLARE_PERSISTANT( EliminationForest )

private:
  friend class constGenericForestIterator;
  friend class constGenericForestPostorderIterator;
  friend class constGenericForestPreorderIterator;

public:
  // Constructors/Destructors
  EliminationForest();
  EliminationForest( const Graph* graph );
  EliminationForest( const Graph* graph, const SharedPtr<PermutationMap>& perm );
  EliminationForest( const Graph* graph, const int n, const int* Parent );
  virtual ~EliminationForest();

private:
  const Graph* g;

public:
  typedef int Key;
  typedef int node;
  typedef constGenericForestIterator const_generic_iterator;
  typedef constGenericForestPostorderIterator const_postorder_iterator;
  typedef constGenericForestPreorderIterator const_preorder_iterator;

  virtual void validate() { GenericForest::validate(); }
  bool isValid() const { return GenericForest::isValid(); }
  void resize(const int n) { g = 0; GenericForest::resize(n); }
  bool reset() { g = 0; return GenericForest::reset(); }
  int size() const { return GenericForest::size(); }
  int queryNTrees() const { return GenericForest::queryNTrees(); }
  int queryParent( const node Child ) const { return GenericForest::queryParent(Child); }
  const SharedArray<int>& getParent() const { return GenericForest::getParent(); }

  // Iterator Access
  const_postorder_iterator begin_postorder() const { return GenericForest::begin_postorder() ; }
  const_postorder_iterator end_postorder() const { return GenericForest::end_postorder(); }
  const_preorder_iterator begin_preorder() const { return GenericForest::begin_preorder(); }
  const_preorder_iterator end_preorder() const { return GenericForest::end_preorder(); }
  const_generic_iterator create_generic_iterator() const { return GenericForest::create_generic_iterator(); }

};

SPINDLE_END_NAMESPACE 

#endif
