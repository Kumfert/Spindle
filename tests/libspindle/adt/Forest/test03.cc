//
// testEliminationForest.cc
//
//  $Id: test03.cc,v 1.2 2000/02/18 01:32:05 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//


#include <iostream.h>
#include <iomanip.h>
#include "spindle/Graph.h"
#include "spindle/EliminationForest.h"
#include "spindle/PermutationMap.h"
 
#include "samples/small_mesh.h"
 
static int mmd_new2old_perm[] = { 0, 2, 6, 8, 1, 7, 3, 4, 5 };
 
#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

int main() {
  ios::sync_with_stdio();
  Graph graph(nvtxs,Aptr,Aind);
  graph.validate();
  
  if ( ! graph.isValid() ) {
    cerr << "Cannot create valid graph" << endl;
  }

  const int n=graph.size();
  // first work on unpermuted graph...
  { 
    EliminationForest etree( &graph );
    etree.validate();
    if (! etree.isValid() ) {
      cerr << "Error: cannot create a valid etree on unpermuted graph." << endl;
      exit(-1);
    }
    cout << "Preorder and Postorder of unpermuted graph." << endl;
    EliminationForest::const_preorder_iterator pre = etree.begin_preorder();
    EliminationForest::const_postorder_iterator post = etree.begin_postorder();
  
    for(int i=0; i<n; ++i ) {
      cout << setw(5) << i << ") " << setw(5) << *pre++ << " " << setw(5) << *post++ << endl;
    }
    cout << endl;
  }

  // now do the same with a permuted graph
  { 
    PermutationMap* p = new PermutationMap(9);
    p->getNew2Old().take( mmd_new2old_perm , 9 );
    p->validate();
    if (! p->isValid() ) {
      cerr << "PermutationMap Object creation failed." << endl;
      exit(-1);
    }
    SharedPtr<PermutationMap> perm( p );
    EliminationForest etree( &graph , perm );
    etree.validate();
    if (! etree.isValid() ) {
      cerr << "Error: cannot create a valid etree on unpermuted graph." << endl;
      exit(-1);
    }
    EliminationForest::const_preorder_iterator pre = etree.begin_preorder();
    EliminationForest::const_postorder_iterator post = etree.begin_postorder();
    cout << "Preorder and Postorder of permuted graph." << endl; 
    for(int i=0; i<n; ++i ) {
      cout << setw(5) << i << ") " << setw(5) << *pre++ // mmd_new2old_perm[*pre++] 
	   << " " << setw(5) << *post++ << endl;
    }
    cout << endl;
  }
}

