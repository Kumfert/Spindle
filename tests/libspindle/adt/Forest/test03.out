Preorder and Postorder of unpermuted graph.
    0)     8     0
    1)     7     1
    2)     6     2
    3)     5     3
    4)     4     4
    5)     3     5
    6)     2     6
    7)     1     7
    8)     0     8

Preorder and Postorder of permuted graph.
    0)     5     0
    1)     4     2
    2)     3     1
    3)     1     6
    4)     0     8
    5)     2     7
    6)     7     3
    7)     6     4
    8)     8     5

